packages/cli/src/databases/repositories/AbstractRepository.ts

```diff
@@ -0,0 +1,65 @@
+import { Repository } from 'typeorm';
+import type { WorkflowEntity } from '../entities/WorkflowEntity';
+import { jsonParse } from 'n8n-workflow';
+import type { JsonValue } from 'n8n-workflow';
+import * as utils from '@/utils';
+import { validate as jsonSchemaValidate } from 'jsonschema';
+import type { Constructor } from '@/Interfaces';
+
+type ParsedJSON = JsonValue;
+
+function mixinQueryMethods<T extends Constructor<{}>>(base: T) {
+	class Derived extends base {
+		static toQueryFilter(rawFilter: string, schema: { properties: object }) {
+			const parsedFilter = jsonParse<ParsedJSON>(rawFilter, {
+				errorMessage: 'Failed to parse JSON into filter object',
+			});
+
+			if (!utils.isObjectLiteral(parsedFilter)) {
+				throw new Error('Parsed filter is not an object');
+			}
+
+			const queryFilter = Object.fromEntries(
+				Object.keys(parsedFilter)
+					.filter((field) => Object.keys(schema.properties).includes(field))
+					.map((field) => [field, parsedFilter[field]]),
+			);
+
+			const fitsSchema = jsonSchemaValidate(queryFilter, schema).valid;
+
+			if (!fitsSchema) throw new Error('Parsed filter object does not fit schema');
+
+			return queryFilter;
+		}
+
+		static toQuerySelect(rawSelect: string, schema: { properties: object }) {
+			const parsedSelect = jsonParse(rawSelect, {
+				errorMessage: 'Failed to parse JSON into select array',
+			});
+
+			if (!utils.isStringArray(parsedSelect)) {
+				throw new Error('Parsed select is not a string array');
+			}
+
+			const keys = Object.keys(schema.properties);
+
+			return parsedSelect.reduce<Record<string, boolean>>((acc, field) => {
+				if (!keys.includes(field)) return acc;
+				return (acc[field] = true), acc;
+			}, {});
+		}
+
+		static toPaginationOptions(rawTake: string, rawSkip?: string) {
+			const MAX_ITEMS = 50;
+
+			const [take, skip] = [rawTake, rawSkip ?? '0'].map((o) => parseInt(o, 10));
+
+			return { skip, take: Math.min(take, MAX_ITEMS) };
+		}
+	}
+
+	return Derived;
+}
+
+/* eslint-disable @typescript-eslint/naming-convention */
+export const BaseWorkflowRepository = mixinQueryMethods(Repository<WorkflowEntity>);
```

Comment:
Maybe call either the class AbstractRepository or the file BaseWorkflowRepository.ts

Reply:
Or neither? :)

I was planning to have here the others as well:

```
export const BaseCredentialsRepository = mixinQueryMethods(Repository<CredentialsEntity>);
export const BaseUsersRepository = mixinQueryMethods(Repository<UserEntity>);
```

Reply:
we can also instead create a new class `AbstractRepository`, add these methods in there, and copy over any missing methods from typeorm's `Repository`, as that code isn't really doing anything but delegating the calls to `this.manager`. typeorm's `Repository` class is an unnecessary abstraction, and I'd be more than happy to stop using it.

Reply:
> copy over any missing methods from typeorm's Repository

To confirm, you mean copying over [all of this](https://github.com/typeorm/typeorm/blob/master/src/repository/Repository.ts) and adding these new methods to the copy? Won't we risk missing any changes by `typeorm` to this class? Seems like a lot just to avoid a little abstraction?

Reply:
I have contributed to that file, and pretty certain that there are no important changes going in there. It's a very dump wrapper on top of the manager. We use less than half of the methods defined in there. I don't think it'd be too bad to copy over the bits we need.
but we don't need to do that right now, as I've been contemplating forking typeorm instead, and merging all these abstractions that we've been building op top of typeorm into a new package.